### PR TITLE
TELEGRAF-2: Data is not sent to output from second named pipe

### PR DESCRIPTION
**Pull request by chayan-04**
Multiple threads were being created since new files function was getting called multiple time. A list is created to check if the called named pipe is already being tailed.

# Required for all PRs

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #

### CODE REVIEW

File: plugins/inputs/tail/tail.go [hunk 1]
```diff
@@ -223,6 +226,21 @@ func (t *Tail) tailNewFiles(fromBeginning bool) error {
...
+				if( e.Value == filepath) {
+					flag = false
+				}
+			}
+
+			if (flag == false) {
+				// we have already started tailing this file
+				continue
+			}
+			l.PushBack(filepath)
```
Comment (dshnayder) [inline plugins/inputs/tail/tail.go:242]: There is a race condition between line 242 and 232. Two threads may add the same file. The both go through loop in line 232, don't find the file in the list, then proceed and both add the file to the list. Pleas find a root cause why the same file gets tailed by more than one thread.

**Review by dshnayder — CHANGES_REQUESTED** [review]

File: plugins/inputs/tail/tail.go [hunk 2]
```diff
@@ -223,6 +226,21 @@ func (t *Tail) tailNewFiles(fromBeginning bool) error {
 				// we're already tailing this file
 				continue
 			}
+
+			//Check whether tailing has started
+			flag = true
```
Comment (dshnayder) [inline plugins/inputs/tail/tail.go:231]: Also, no point of checking if file is tailed, as there is exactly the same check on line 225

File: plugins/inputs/tail/tail.go [hunk 3]
```diff
@@ -223,6 +226,21 @@ func (t *Tail) tailNewFiles(fromBeginning bool) error {
 				// we're already tailing this file
```
Comment (dshnayder) [inline plugins/inputs/tail/tail.go:226]: I'd suggest to print `file` here, to check if the same pipe is tailed ore than once.

Comment (chayan-04) [reply plugins/inputs/tail/tail.go:226]: > I'd suggest to print `file` here, to check if the same pipe is tailed ore than once.

@dshnayder , this keeps on flooding telegraf log as tailNewFiles function is keep on getting called multiple times

File: plugins/inputs/tail/tail.go (hunk 1 above)
Comment (chayan-04) [reply plugins/inputs/tail/tail.go:242]: > There is a race condition between line 242 and 232. Two threads may add the same file. The both go through loop in line 232, don't find the file in the list, then proceed and both add the file to the list. Pleas find a root cause why the same file gets tailed by more than one thread.

It wont race, we are adding the entry to list only if that same entry is not already there in list. this is to make sure only one thread handles one named pipe.

Comment (dshnayder) [reply plugins/inputs/tail/tail.go:242]: Imaging how two thread execute this code at the same time at the same lines. This is a classic race condition when you check if object is not added, then add it a few lines later.

File: plugins/inputs/tail/tail.go (hunk 3 above)
Comment (dshnayder) [reply plugins/inputs/tail/tail.go:226]: How this function gets called multiple times? It supposed to be called once pr pipe, so its two times for you config.

File: plugins/inputs/tail/tail.go (hunk 1 above)
Comment (chayan-04) [reply plugins/inputs/tail/tail.go:242]: > Imaging how two thread execute this code at the same time at the same lines. This is a classic race condition when you check if object is not added, then add it a few lines later.

@dshnayder 2 threads wont reach this line with same filename

File: plugins/inputs/tail/tail.go (hunk 3 above)
Comment (chayan-04) [reply plugins/inputs/tail/tail.go:226]: > How this function gets called multiple times? It supposed to be called once pr pipe, so its two times for you config.

@dshnayder , It keeps on getting called multiple times throughout telegraf lifetime. let me investigate on how. It does not gets called only twice. thats why we have line 225 and line 235 to check if we have already started adding the tail.

**Pull request merged**